README.md
# Saxophone Fingering Chart Generator   

This program takes a MIDI file as input and generates fingering charts for saxophones matching the notes in that midi file. Time, note length, and other musical elements are not preserved, this is just a way to learn the fingering for each note.
That said, this may expand and become fancier down the road.

Fingerings are decided by a simple weighted algorithm that picks fingerings for a note based the keys involved in the previous note. There's probably a smarter way to do this, I'm open to ideas.

## Usage   
- fingering_chart [midi_file...]

Just open a midi file with `fingering_chart`, and wait for charts to generate! You can also run it via the command-line by specifying one or more midi files as shown above. Make sure the `cfg.ron` file is in the same directory as `fingering_chart`. See below for additional settings.

### cfg.ron
cfg.ron is the configuration file:
- `transposition_type` Sets the transposition. May use `Alto`, `Tenor`, `Baritone`, `Soprano`, `Bass`, `CMelody`, `Contrabass`, `Sopranino`, `Subcontrabass`, and `Sopranissimo`.
- `output_path` Sets the output path. If this is a non-existent directory, will attempt to create.
- `output_format` Sets the output format. May use `Tracks`, `Rows`, or `Separate`. `Tracks` outputs charts for each midi track as a single file. `Rows` splits tracks up into multiple files each containing a horizontal row of notes. `Separate` outputs each individual note as a separate file.
- `spacing` Sets the visual spacing between consecutive notes. Not applicable when using `Separate` output format.
- `notes_per_row` Sets the number of notes per row. Not applicable when using `Separate` output format.
- `notes` You can now fully customize what notes are supported and what fingerings each note will use. Detailed instructions exist inside the cfg.ron file.

## Planned Features   
- Fingerings for altissimo/trill
- Detect if a midi file has notes out of range, have an option to automatically attempt to transpose the midi file to fit in the sax range (by octave first, then by semitone). Currently notes out of range are just left out with a warning.

## Example   
![The Lizards](https://raw.githubusercontent.com/Eolu/fingering_chart/master/examples/lizards.png)   


CHANGELOG.md
# Changelog

## [0.2.5] - 3/21/2022

- Fixed a bug causing certain transpoisition types to fail.

## [0.2.4] - 10/21/2021

- Image data is now generated in a much more sane way through build scripts.
- Implemented a workaround for a bug in which for some reason some midi files parse with 2 of every note. (Is this is bug in the midly library?)

## [0.2.1] - 9/5/2020

- Realized charts were generating at 50% size right after publishing 2.0. Here's the fix.

## [0.2.0] - 9/5/2020

- Charts arte now much less busy, opting to not show notes when they aren't relevant to a fingering. The octave key and 6 front keys are always shown.
- Fingering charts are now 100% precedurally generated. No more archives of chart files! Releases will be a single executable with a config file.
- Got rid of a silly note-loading macro that was almost certainly a bad idea from the start. Now using the magic of ron and serde. You can edit the configuration file to add more notes as well as modify the fingering of existing notes!

#### Future plans:
- Still gives up if config file is not present, considering solutions (eg generating a default if it's not there).

## [0.1.3] - 8/29/2020

- Fingering decision logic slightly simplified.
- More refactoring and began adding some tests.
- Added some useful documention inside keys.cfg.

## [0.1.2] - 8/19/2020

- Added logic which tries to make vaguely intelligent decisions about which fingering chart to pick. It checks the fingering before and after,
and tries to pick one with more keys in common to its neighbors.
- To support the above change, added another configuration file that gives some control over notes. Not well-documented yet, but that's coming soon.
- Code refactoring to handle the increasing size of this project.

#### Future plans:

- Considering generating key images at runtime instead of storing them in a bunch of images. It would be fairly easy to do so now that notes store key information.
- Altissimo/Trill fingerings are looking like good ideas for the next update, although a bit more infrastructure may be needed.

## [0.1.0] - 8/14/2020

Initial release.

It doesn't contain all of the intended features, but it works! The worst feature of this release to look out for is that when using a midi file with notes outside the supported range, those notes will left out of the charts at the end. A warning is printed to the console, but even that may be missed. Therefore, make sure you only try to convert songs with notes ranging from Bb2 to F5 (slightly higher or lower when taking transposition into account)!

src/tests/mod.rs
const OUTPUT_DIR: &str = "test_out";

/// Prints out each note image
#[test]
fn output_notes()
{
    std::fs::create_dir_all(OUTPUT_DIR).expect("Failed to create test output dir");
    for note in (46..78).filter_map(crate::note::Note::get)
    {
        for fingering in &note.fingerings
        {
            let outfile = format!("{}/{}_{}.png", OUTPUT_DIR, note.byte, fingering.keys.as_u32());
            println!("Writing {:?} to {}", fingering.keys, outfile);
            fingering.image.save(outfile).expect("Failed to save note image");
        }
    };
}

/// Prints the bits that make up other images
#[test]
pub fn output_images()
{
    use crate::keys::image_data::*;
    use std::fs::*;

    std::fs::create_dir_all("./test_image_data").expect("Failed to create test output dir");

    macro_rules! write_images
    {
        ($($name:ident),*) => 
        {
            $(
                let mut file = File::create(concat!("test_image_data/", stringify!($name), ".png")).unwrap();
                $name.write_to(&mut file, image::ImageOutputFormat::Png).unwrap();
            )*
        }
    }

    write_images!
    (
        SEPARATOR, A, B, C, D, E, F, G, FLAT, FRONT_KEY, FRONT_KEY_OFF,
        FRONT_F_KEY, FRONT_F_KEY_OFF,PALM_KEY, PALM_KEY_OFF, OCTAVE_KEY, 
        OCTAVE_KEY_OFF, BIS_KEY, BIS_KEY_OFF, LOW_A_KEY, LOW_A_KEY_OFF, 
        SIDE_KEY, SIDE_KEY_OFF, HIGH_F_SHARP_KEY, HIGH_F_SHARP_KEY_OFF, 
        F_SHARP_KEY, F_SHARP_KEY_OFF, LOW_E_FLAT_KEY, LOW_E_FLAT_KEY_OFF, 
        LOW_C_KEY, LOW_C_KEY_OFF, G_SHARP_KEY, G_SHARP_KEY_OFF, SMALL_PINKY_KEY, 
        SMALL_PINKY_KEY_OFF, LOW_B_FLAT_KEY, LOW_B_FLAT_KEY_OFF
    );
}

src/main.rs

#[macro_use]
extern crate lazy_static;

#[macro_use]
mod note;
mod keys;
mod track;
mod song;

#[cfg(test)]
mod tests;

use song::*;
use ron::de::from_str;
use serde::Deserialize;
use std::{env, path::Path, collections::BTreeMap};
use image::{error::ImageError};
use enumset::*;
use keys::Key;

// Load configuration before executing main program
lazy_static!
{
    static ref CONFIG: Result<Config, ron::error::Error> = 
    std::fs::read_to_string("./cfg.ron")
        .map_err(ron::error::Error::from)
        .and_then(|str| from_str(&str));
}

/// The note transposition to use, supports saxes of any kind
#[derive(Copy, Clone, Deserialize)]
pub enum TranspositionType
{
    Sopranissimo = -22,
    Sopranino = -15,
    Soprano = -10,
    Alto = -3,
    CMelody = 0,
    Tenor = 2,
    Baritone = 9,
    Bass = 14,
    Contrabass = 21,
    Subcontrabass = 26
}

/// The output format, determines how the chart images are layed out.
#[derive(Copy, Clone, Deserialize, PartialEq)]
pub enum OutputFormat
{
    Separate,
    Rows,
    Tracks
}

/// Data from the loaded cfg.ron file.
#[derive(Deserialize)]
pub struct Config 
{
    transposition_type: TranspositionType,
    output_path: String,
    output_format: OutputFormat,
    spacing: usize,
    notes_per_row: usize,
    notes: BTreeMap<u8, Vec<EnumSet<Key>>> 
}

/// Entry-point
fn main() -> Result<(), ImageError>
{
    match &*CONFIG
    {
        Ok(config) =>
        {
            for midi_file in env::args().skip(1)
            {
                let fingering_chart = Song::load(&midi_file, config.transposition_type as i16);
                let dir_name = Path::new(&midi_file).file_stem().unwrap().to_string_lossy();
                let output_path = format!("{}/{}", config.output_path, dir_name);
                match config.output_format
                {
                    OutputFormat::Tracks => fingering_chart.output_entire(&output_path, config.notes_per_row, config.spacing)?,
                    OutputFormat::Rows => fingering_chart.output_rows(&output_path, config.notes_per_row, config.spacing)?,
                    OutputFormat::Separate => fingering_chart.output_cells(&output_path)?
                }
            }
        }
        Err(e) => eprintln!("Failed to load config: {}", e)
    }
    Ok(())
}

src/note.rs
use enumset::EnumSet;
use std::collections::HashMap;
use crate::keys::Key;
use crate::CONFIG;

/// Type alias for note constants
pub type NoteConst = &'static Note;

/// Struct used for individual notes
pub struct Note
{
    pub byte: u8,
    pub fingerings: Vec<Fingering>
}

impl PartialEq for Note
{
    fn eq(&self, other: &Self) -> bool 
    {
        self.byte == other.byte && self.fingerings == other.fingerings
    }
}

/// Struct used for individual fingering charts, may be multiple per note
pub struct Fingering
{
    pub keys: EnumSet<Key>,
    pub image: image::DynamicImage
}

impl PartialEq for Fingering
{
    fn eq(&self, other: &Self) -> bool 
    {
        self.keys == other.keys
    }
}

// Define notes and load note images using lazy_static
lazy_static!
{
    /// Map of loaded note constants
    pub static ref NOTES: HashMap<u8, Note> = 
    {
        match &*CONFIG
        {
            Ok(config) => 
            {
                let mut notes: HashMap<u8, Note> = HashMap::new();
                for (byte, fingerings) in &config.notes
                {
                    let fingerings = fingerings
                        .into_iter()
                        .map(|&enumset| Fingering::new(enumset, *byte))
                        .collect();
                    match notes.get_mut(&byte)
                    {
                        None => 
                        {
                            notes.insert(*byte, Note { byte: *byte, fingerings });
                        },
                        Some(note) => 
                        {
                            note.fingerings.extend(fingerings);
                        }
                    };
                }
                notes
            },
            Err(e) => 
            {
                eprintln!("Failed to load config: {}", e);
                HashMap::new()
            }
        }
    };
}

impl Note
{
    /// Access a note via it's midi byte index.
    pub fn get(byte: u8) -> Option<NoteConst>
    {
        NOTES.get(&byte)
    }
}

impl Fingering
{
    /// Fingering contructor
    pub fn new(keys: EnumSet<Key>, byte: u8) -> Fingering
    {
        let mut image = Fingering::gen_chart(keys);
        Note::include_note_name(byte, &mut image);
        Fingering { keys, image }
    }
}

src/song.rs
use crate::{track::*, note::{Note, NoteConst}};
use std::{fs, collections::HashSet};
use midly::{Smf, TrackEventKind::*, MidiMessage::*};
use image::{error::ImageError, imageops::FilterType};

/// Entire song, just a vec of tracks with some methods
pub struct Song(Vec<Track>);

impl Song
{
    /// Parse a midi file. Generate a list containing all tracks. Tracks themselves are simply lists of notes.
    pub fn load(midi_path: &str, transposition: i16) -> Song
    {
        // Parse a midi file
        let raw_data = fs::read(midi_path).expect("Failed to load midi data");
        let midi = Smf::parse(&raw_data).expect("Failed to parse midi data");

        // Keep track of out-of-range notes
        let mut out_of_range: HashSet<u8> = HashSet::new();

        // Iterate through the midi file and collect notes
        let mut tracks: Vec<Track> = midi.tracks
            .iter()
            .map(|track| track
                .iter()
                .filter_map(|event| 
                    if let Midi { channel: _, message: NoteOn {key, vel: _} } = event.kind 
                    { 
                        let true_key = (key.as_int() as i16 + transposition) as u8;
                        let candidates = Note::get(true_key);
                        match candidates
                        {
                            None => 
                            {
                                if out_of_range.insert(true_key)
                                {
                                    eprintln!("Note out of range: {}", true_key);
                                }
                                None
                            },
                            note => note
                        }
                    } 
                    else 
                    { 
                        None 
                    })
                .collect::<Vec<NoteConst>>())
            .filter_map(|candidates| 
                if candidates.is_empty() 
                { 
                    None 
                } 
                else 
                {
                    Some(Track(candidates)) 
                })
            .collect();
        // Determine if every note in track is a duplicate
        // TODO: This should not be necessary, find out why this happens
        for track in tracks.iter_mut()
        {
            if track.notes().chunks(2).filter(|chunk| (chunk.len() == 2 && chunk[0] != chunk[1]) || chunk.len() == 1).next().is_none()
            {
                for i in (0..track.notes().len()).rev()
                {
                    if i % 2 == 0
                    {
                        track.notes_mut().remove(i);
                    }
                }
                
            }
        }
        Song(tracks)
    }

    /// Output chart with each cell as an individual file
    pub fn output_cells(&self, output_path: &str) -> Result<(), ImageError>
    {
        for (i, track) in self.tracks().enumerate()
        {
            let track_path = format!("{}/track{}", output_path, i);
            fs::create_dir_all(&track_path)?;
            for (cell, image) in track.cell_images().iter().enumerate()
            {
                let image = image.resize(image.width() * 2, image.height() * 2, FilterType::Nearest);
                image.save(format!("{}/{}.png", track_path, cell))?;
            }
        }
        Ok(())
    }

    /// Output chart with each row as an individual file
    pub fn output_rows(&self, output_path: &str, notes_per_row: usize, spacing: usize) -> Result<(), ImageError>
    {
        for (i, track) in self.tracks().enumerate()
        {
            let track_path = format!("{}/track{}", output_path, i);
            fs::create_dir_all(&track_path)?;
            for (row, image) in track.row_images(notes_per_row, spacing).iter().enumerate() 
            {
                let image = image.resize(image.width() * 2, image.height() * 2, FilterType::Nearest);
                image.save(format!("{}/row{}.png", track_path, row))?;
            }
        }
        Ok(())
    }

    /// Generate a chart and output to the given directory.
    pub fn output_entire(&self, output_path: &str, notes_per_row: usize, spacing: usize) -> Result<(), ImageError>
    {
        fs::create_dir_all(output_path)?;
        for (i, track) in self.tracks().enumerate()
        {
            let image = track.track_image(notes_per_row, spacing);
            let image = image.resize(image.width() * 2, image.height() * 2, FilterType::Nearest);
            image.save(format!("{}/track{}.png", output_path, i))?;
        }
        Ok(())
    }

    /// Access notes vector
    fn tracks(&self) -> impl Iterator<Item = &Track>
    {
        self.0.iter()
    }
}



res/build.rs
use std::path::Path;
use const_gen::*;

#[cfg(windows)]
extern crate winres;

#[cfg(windows)]
fn main() 
{
    let mut res = winres::WindowsResource::new();
    res.set_icon("res/icon.ico");
    res.compile().unwrap();
    build();
}

#[cfg(unix)]
fn main() 
{
    build();
}

macro_rules! load_images
{
    ($($name:ident),*) => 
    {
        vec!
        [
            $(
                const_declaration!(pub $name = load_image(concat!(stringify!($name), ".png")))
            ),*
        ].join("\n")
    }
}

fn build()
{
    let out_dir = std::env::var_os("OUT_DIR").unwrap();
    let dest_path = Path::new(&out_dir).join("raw_image_data.rs");

    let const_decs = load_images!
    {
        SEPARATOR, A, B, C, D, E, F, G, FLAT, FRONT_KEY, FRONT_KEY_OFF,
        FRONT_F_KEY, FRONT_F_KEY_OFF,PALM_KEY, PALM_KEY_OFF, OCTAVE_KEY, 
        OCTAVE_KEY_OFF, BIS_KEY, BIS_KEY_OFF, LOW_A_KEY, LOW_A_KEY_OFF, 
        SIDE_KEY, SIDE_KEY_OFF, HIGH_F_SHARP_KEY, HIGH_F_SHARP_KEY_OFF, 
        F_SHARP_KEY, F_SHARP_KEY_OFF, LOW_E_FLAT_KEY, LOW_E_FLAT_KEY_OFF, 
        LOW_C_KEY, LOW_C_KEY_OFF, G_SHARP_KEY, G_SHARP_KEY_OFF, SMALL_PINKY_KEY, 
        SMALL_PINKY_KEY_OFF, LOW_B_FLAT_KEY, LOW_B_FLAT_KEY_OFF
    };

    std::fs::write(&dest_path, const_decs).unwrap();
}

fn load_image<'a>(name: &str) -> Vec<u8>
{
    let root = std::env::var("CARGO_MANIFEST_DIR").unwrap();
    std::fs::read(Path::new(&root).join("image_data").join(name))
        .expect(&format!("Failed to read image file: {}", name))
}



src/track.rs
use crate::note::{NoteConst, Fingering};
use image::{DynamicImage, GenericImage};

/// Struct used for tracks
pub struct Track(pub Vec<NoteConst>);

impl Track
{
    /// Generates an image for the entire track
    pub fn track_image(&self, notes_per_row: usize, spacing: usize) -> DynamicImage
    {
        let mut track_image: DynamicImage = DynamicImage::new_rgb8(0, spacing as u32);
        for row_image in self.row_images(notes_per_row, spacing)
        {
            let previous = track_image;
            let new_width = std::cmp::max(previous.width(), row_image.width());
            let new_height = previous.height() + row_image.height() + spacing as u32;
            track_image = DynamicImage::new_rgb8(new_width, new_height);
            track_image.copy_from(&previous, 0, 0).expect("Failed to copy track image");
            track_image.copy_from(&row_image, 0, previous.height()).expect("Failed to generate track image");
        }
        track_image
    }

    /// Generates images in rows
    pub fn row_images(&self, notes_per_row: usize, spacing: usize) -> Vec<DynamicImage>
    {
        let fingerings = self.select_fingerings();
        (0..fingerings.len())
            .take_while(|row| row * notes_per_row < self.notes().len())
            .map(|row|
            {
                let mut row_image: DynamicImage = DynamicImage::new_rgb8(spacing as u32, 0);
                for image in (0..notes_per_row)
                    .take_while(|col| col + (row * notes_per_row) < self.notes().len())
                    .map(|col| fingerings[(row * notes_per_row) + col])
                {
                    let previous = row_image;
                    let new_width = previous.width() + image.width() + spacing as u32;
                    let new_height = std::cmp::max(previous.height(), image.height());
                    row_image = DynamicImage::new_rgb8(new_width, new_height);
                    row_image.copy_from(&previous, 0, 0).expect("Failed to copy row image");
                    row_image.copy_from(image, previous.width(), 0).expect("Failed to generate row image");
                }
                row_image
            }).collect()
    }

    /// Returns an iterator into cell images. The images themselves are 
    /// generated at load-time, so this method is low-cost.
    pub fn cell_images(&self) -> Vec<&DynamicImage>
    {
        self.select_fingerings()
    }

    /// Access notes vector
    pub fn notes(&self) -> &Vec<NoteConst>
    {
        &self.0
    }

    /// Access notes vector
    pub fn notes_mut(&mut self) -> &mut Vec<NoteConst>
    {
        &mut self.0
    }

    /// Go through candidates and determine which fingerings to use
    fn select_fingerings(&self) -> Vec<&DynamicImage>
    {
        let mut previous_choice: Option<&Fingering> = None;
        self.notes()
            .iter()
            .enumerate()
            .map(|(i, notes)| (i, &notes.fingerings))
            .filter_map(|(i, fingerings)| 
            {
                // Determine how alike a note is from a list of candidates
                fn intersection<'a>(note: &Fingering, candidates: impl Iterator<Item = &'a Fingering>) -> i32
                {
                    candidates
                        .map(|candidate|candidate.keys)
                        .map(|keys|note.keys.intersection(keys).len())
                        .max()
                        .unwrap() as i32
                }

                // Determine how different a note is from a list of candidates
                fn difference<'a>(note: &Fingering, candidates: impl Iterator<Item = &'a Fingering>) -> i32
                {
                    candidates
                        .map(|candidate|candidate.keys)
                        .map(|keys|note.keys.symmetrical_difference(keys).len())
                        .min()
                        .unwrap() as i32
                }

                // Take two candidates notes and one or more others to compare against, determine which note is closer.
                // Note 1 is favored slightly in that it will be used in the case of equivalence
                fn compare_notes<'a>(note_1: &'a Fingering, note_2: &'a Fingering, siblings: &Vec<&Fingering>) -> &'a Fingering
                {
                    let note_1_intersection = intersection(note_1, siblings.iter().map(|f|*f));
                    let note_1_diff = difference(note_1, siblings.iter().map(|f|*f));
                    let note_2_intersection = intersection(note_2, siblings.iter().map(|f|*f));
                    let note_2_diff = difference(note_2, siblings.iter().map(|f|*f));
                    if note_2_intersection - note_2_diff > note_1_intersection - note_1_diff
                    {
                        note_2
                    }
                    else
                    {
                        note_1
                    }
                }

                // Determine what fingering to actually use
                previous_choice = Some(if fingerings.len() == 1
                {
                    &fingerings[0]
                }
                else if i == 0
                {
                    let siblings = self.notes()[i + 1].fingerings.iter().collect();
                    fingerings.iter().fold(&fingerings[0], |note_1, note_2| &compare_notes(note_1, note_2, &siblings))
                }
                else
                {
                    let sibling = vec!(previous_choice.unwrap());
                    fingerings.iter().fold(&fingerings[0], |note_1, note_2| &compare_notes(note_1, note_2, &sibling))
                });

                // Return the choice made above (it's actually the "current_choice" at this point)
                previous_choice
            })
            .map(|fingering| &fingering.image)
            .collect()
    }
}


src/keys/mod.rs
pub(crate) mod image_data;

use crate::note::{Note, Fingering};
use image_data::*;
use enumset::*;
use image::{DynamicImage, GenericImage, GenericImageView};
use serde::{Serialize, Deserialize};

#[derive(EnumSetType, Debug, Deserialize, Serialize)]
#[enumset(serialize_as_list)]
pub enum Key
{
    Octave,
    Left1,
    Left2,
    Left3,
    FrontF,
    Bis,
    PalmD,
    PalmEflat,
    PalmF,
    Gsharp,
    LowCsharp,
    LowB,
    LowBflat,
    Right1,
    Right2,
    Right3,
    Fsharp,
    SideE,
    SideC,
    SideBis,
    HighFsharp,
    LowEflat,
    LowC,
    // Baritone only
    LowA
}

impl Fingering
{
    /// Fingering charts are generated here. Note names must be added after
    pub fn gen_chart(keys: EnumSet<Key>) -> DynamicImage
    {
        let mut chart = gen_base_chart();

        // Depending on which keys are visible, show more greyed-out keys
        if !keys.is_disjoint(Key::LowEflat | Key::LowC)
        {
            Key::LowEflat.include_key_image(false, &mut chart);
            Key::LowC.include_key_image(false, &mut chart);
        }
        if !keys.is_disjoint(Key::Gsharp | Key::LowCsharp | Key::LowB | Key::LowBflat)
        {
            Key::Gsharp.include_key_image(false, &mut chart);
            Key::LowCsharp.include_key_image(false, &mut chart);
            Key::LowB.include_key_image(false, &mut chart);
            Key::LowBflat.include_key_image(false, &mut chart);
        }
        if !keys.is_disjoint(Key::PalmD | Key::PalmEflat | Key::PalmF)
        {
            Key::PalmD.include_key_image(false, &mut chart);
            Key::PalmEflat.include_key_image(false, &mut chart);
            Key::PalmF.include_key_image(false, &mut chart);
        }
        if !keys.is_disjoint(Key::SideE | Key::SideC | Key::SideBis | Key::Fsharp | Key::HighFsharp)
        {
            Key::SideE.include_key_image(false, &mut chart);
            Key::SideC.include_key_image(false, &mut chart);
            Key::SideBis.include_key_image(false, &mut chart);
        }

        for key in keys
        {
            key.include_key_image(true, &mut chart);
        }
        chart
    }
}

impl Note
{
    pub fn include_note_name(byte: u8, image: &mut DynamicImage)
    {
        let (color, name, flat) = Note::get_image_data(byte);
        for (x, y, inner_color) in name.pixels()
        {
            if inner_color == BLACK
            {
                let x = x + NAME_LOCATION.0 as u32;
                let y = y + NAME_LOCATION.1 as u32;
                image.put_pixel(x, y, color)
            }
        }
        if let Some(flat) = flat
        {
            for (x, y, inner_color) in flat.pixels()
            {
                if inner_color == BLACK
                {
                    let x = x + FLAT_LOCATION.0 as u32;
                    let y = y + FLAT_LOCATION.1 as u32;
                    image.put_pixel(x, y, color)
                }
            }
        }
    }
}

impl Key
{
    /// Include this key image on a chart
    fn include_key_image(&self, filled: bool, image: &mut DynamicImage)
    {
        let (x, y, note_data, note_off_data) = self.get_image_data();
        image.copy_from(if filled {note_data} else {note_off_data}, x as u32, y as u32)
            .expect("Failed to generate key image");
    }
}

/// Generates a blank fingering chart
fn gen_base_chart() -> DynamicImage
{
    let mut image: DynamicImage = DynamicImage::new_rgba8(CHART_SIZE.0 as u32, CHART_SIZE.1 as u32);
    image.invert();

    // All charts start with 6 unfilled notes and a separator
    Key::Octave.include_key_image(false, &mut image);
    Key::Left1.include_key_image(false, &mut image);
    Key::Left2.include_key_image(false, &mut image);
    Key::Left3.include_key_image(false, &mut image);
    Key::Right1.include_key_image(false, &mut image);
    Key::Right2.include_key_image(false, &mut image);
    Key::Right3.include_key_image(false, &mut image);
    image.copy_from(&*SEPARATOR, SEP_LOCATION.0 as u32, SEP_LOCATION.1 as u32)
        .expect("Failed to generate separator image");
    image
}



src/keys/image_data.rs
use super::Key;
use crate::note::Note;
use image::{DynamicImage, Rgba};

// Various color constants used in chart generation
pub const BLACK: Rgba::<u8> = Rgba([0,0,0,255]);
pub const OCTAVE_2_COLOR: Rgba::<u8> = Rgba([164,120,59,255]);
pub const OCTAVE_3_COLOR: Rgba::<u8> = Rgba([97,90,199,255]);
pub const OCTAVE_4_COLOR: Rgba::<u8> = Rgba([184,94,191,255]);
pub const OCTAVE_5_COLOR: Rgba::<u8> = Rgba([76,158,91,255]);

/// The size of charts, in pixels. The actual charts output may multiply this by a whole-number factor
pub const CHART_SIZE: (u8,u8) = (63, 118);
/// The pixel location of the note-name on the chart
pub const NAME_LOCATION: (u8,u8) = (41,104);
/// The pixel location of the flat symbol that goes next to the note-name
pub const FLAT_LOCATION: (u8,u8) = (51,104);
/// The pixel location of horizontal separator between front keys
pub const SEP_LOCATION: (u8,u8) = (20,65);

/// This module is a container for literal raw png image data
mod raw_image_data
{
    include!(concat!(env!("OUT_DIR"), "/raw_image_data.rs"));
}

// Generated images.
lazy_static!
{
    pub static ref SEPARATOR: DynamicImage = load_image(raw_image_data::SEPARATOR);
    pub static ref A: DynamicImage = load_image(raw_image_data::A);
    pub static ref B: DynamicImage = load_image(raw_image_data::B);
    pub static ref C: DynamicImage = load_image(raw_image_data::C);
    pub static ref D: DynamicImage = load_image(raw_image_data::D);
    pub static ref E: DynamicImage = load_image(raw_image_data::E);
    pub static ref F: DynamicImage = load_image(raw_image_data::F);
    pub static ref G: DynamicImage = load_image(raw_image_data::G);
    pub static ref FLAT: DynamicImage = load_image(raw_image_data::FLAT);
    pub static ref FRONT_KEY: DynamicImage = load_image(raw_image_data::FRONT_KEY);
    pub static ref FRONT_KEY_OFF: DynamicImage = load_image(raw_image_data::FRONT_KEY_OFF);
    pub static ref FRONT_F_KEY: DynamicImage = load_image(raw_image_data::FRONT_F_KEY);
    pub static ref FRONT_F_KEY_OFF: DynamicImage = load_image(raw_image_data::FRONT_F_KEY_OFF);
    pub static ref PALM_KEY: DynamicImage = load_image(raw_image_data::PALM_KEY);
    pub static ref PALM_KEY_OFF: DynamicImage = load_image(raw_image_data::PALM_KEY_OFF);
    pub static ref OCTAVE_KEY: DynamicImage = load_image(raw_image_data::OCTAVE_KEY);
    pub static ref OCTAVE_KEY_OFF: DynamicImage = load_image(raw_image_data::OCTAVE_KEY_OFF);
    pub static ref BIS_KEY: DynamicImage = load_image(raw_image_data::BIS_KEY);
    pub static ref BIS_KEY_OFF: DynamicImage = load_image(raw_image_data::BIS_KEY_OFF);
    pub static ref LOW_A_KEY: DynamicImage = load_image(raw_image_data::LOW_A_KEY);
    pub static ref LOW_A_KEY_OFF: DynamicImage = load_image(raw_image_data::LOW_A_KEY_OFF);
    pub static ref SIDE_KEY: DynamicImage = load_image(raw_image_data::SIDE_KEY);
    pub static ref SIDE_KEY_OFF: DynamicImage = load_image(raw_image_data::SIDE_KEY_OFF);
    pub static ref HIGH_F_SHARP_KEY: DynamicImage = load_image(raw_image_data::HIGH_F_SHARP_KEY);
    pub static ref HIGH_F_SHARP_KEY_OFF: DynamicImage = load_image(raw_image_data::HIGH_F_SHARP_KEY_OFF);
    pub static ref F_SHARP_KEY: DynamicImage = load_image(raw_image_data::F_SHARP_KEY);
    pub static ref F_SHARP_KEY_OFF: DynamicImage = load_image(raw_image_data::F_SHARP_KEY_OFF);
    pub static ref LOW_E_FLAT_KEY: DynamicImage = load_image(raw_image_data::LOW_E_FLAT_KEY);
    pub static ref LOW_E_FLAT_KEY_OFF: DynamicImage = load_image(raw_image_data::LOW_E_FLAT_KEY_OFF);
    pub static ref LOW_C_KEY: DynamicImage = load_image(raw_image_data::LOW_C_KEY);
    pub static ref LOW_C_KEY_OFF: DynamicImage = load_image(raw_image_data::LOW_C_KEY_OFF);
    pub static ref G_SHARP_KEY: DynamicImage = load_image(raw_image_data::G_SHARP_KEY);
    pub static ref G_SHARP_KEY_OFF: DynamicImage = load_image(raw_image_data::G_SHARP_KEY_OFF);
    pub static ref SMALL_PINKY_KEY: DynamicImage = load_image(raw_image_data::SMALL_PINKY_KEY);
    pub static ref SMALL_PINKY_KEY_OFF: DynamicImage = load_image(raw_image_data::SMALL_PINKY_KEY_OFF);
    pub static ref LOW_B_FLAT_KEY: DynamicImage = load_image(raw_image_data::LOW_B_FLAT_KEY);
    pub static ref LOW_B_FLAT_KEY_OFF: DynamicImage = load_image(raw_image_data::LOW_B_FLAT_KEY_OFF);
}

impl Note
{
    /// Gets the pixel location of a keyname on the fingering chart and the image that goes there
    pub fn get_image_data<'a>(byte: u8) -> (Rgba::<u8>, &'a DynamicImage, Option<&'a DynamicImage>)
    {
        let (name, flat) = match byte % 12
        {
            0 => (&*C, None),
            1 => (&*C, Some(&*FLAT)),
            2 => (&*D, None),
            3 => (&*D, Some(&*FLAT)),
            4 => (&*E, None),
            5 => (&*F, None),
            6 => (&*F, Some(&*FLAT)),
            7 => (&*G, None),
            8 => (&*G, Some(&*FLAT)),
            9 => (&*A, None),
            10 => (&*A, Some(&*FLAT)),
            11 => (&*B, None),
            n => panic!("Modulus returned impossible number: {}", n)
        };
        let color = match byte
        {
            n if n < 48 => OCTAVE_2_COLOR,
            n if n < 60 => OCTAVE_3_COLOR,
            n if n < 72 => OCTAVE_4_COLOR,
            _ => OCTAVE_5_COLOR
        };
        return (color, name, flat);
    }
}

impl Key
{
    /// Gets the pixel location of a key on the fingering chart and the image that goes there
    pub fn get_image_data(&self) -> (u8, u8, &'static DynamicImage, &'static DynamicImage)
    {
        match *self
        {
            Self::Octave => (1,12,&OCTAVE_KEY,&OCTAVE_KEY_OFF),
            Self::Left1 => (22,23,&FRONT_KEY,&FRONT_KEY_OFF),
            Self::Left2 => (22,38,&FRONT_KEY,&FRONT_KEY_OFF),
            Self::Left3 => (22,53,&FRONT_KEY,&FRONT_KEY_OFF),
            Self::FrontF => (22,15,&FRONT_F_KEY,&FRONT_F_KEY_OFF),
            Self::Bis => (29,31,&BIS_KEY,&BIS_KEY_OFF),
            Self::PalmD => (53,10,&PALM_KEY,&PALM_KEY_OFF),
            Self::PalmEflat => (44,1,&PALM_KEY,&PALM_KEY_OFF),
            Self::PalmF => (37,14,&PALM_KEY,&PALM_KEY_OFF),
            Self::Gsharp => (43,61,&G_SHARP_KEY,&G_SHARP_KEY_OFF),
            Self::LowCsharp => (51,69,&SMALL_PINKY_KEY,&SMALL_PINKY_KEY_OFF),
            Self::LowB => (43,69,&SMALL_PINKY_KEY,&SMALL_PINKY_KEY_OFF),
            Self::LowBflat => (43,75,&LOW_B_FLAT_KEY,&LOW_B_FLAT_KEY_OFF),
            Self::Right1 => (22,68,&FRONT_KEY,&FRONT_KEY_OFF),
            Self::Right2 => (22,83,&FRONT_KEY,&FRONT_KEY_OFF),
            Self::Right3 => (22,98,&FRONT_KEY,&FRONT_KEY_OFF),
            Self::Fsharp => (9,93,&F_SHARP_KEY,&F_SHARP_KEY_OFF),
            Self::SideE => (7,62,&SIDE_KEY,&SIDE_KEY_OFF),
            Self::SideC => (7,71,&SIDE_KEY,&SIDE_KEY_OFF),
            Self::SideBis => (7,80,&SIDE_KEY,&SIDE_KEY_OFF),
            Self::HighFsharp => (12,84,&HIGH_F_SHARP_KEY,&HIGH_F_SHARP_KEY_OFF),
            Self::LowEflat => (4,105,&LOW_E_FLAT_KEY,&LOW_E_FLAT_KEY_OFF),
            Self::LowC => (4,110,&LOW_C_KEY,&LOW_C_KEY_OFF),
            Self::LowA => (3,34,&LOW_A_KEY,&LOW_A_KEY_OFF)
        }
    }
}

/// Load image, if this fails we might as well panic
fn load_image(raw_image: &[u8]) -> DynamicImage
{
    image::load_from_memory_with_format(raw_image, image::ImageFormat::Png)
        .expect("Failed to generate image data")
}
